### DISCUSSION TITLE
Use of the generated isTYPE() function

### DISCUSSION THREAD

**Discussion by goto40** [eclipse-langium/langium#617] [Q&A] [answered]
I am using langium 0.4.0.

I created a minimal language which models structs, which contain a list of attributes. These attributes can be of two types: arrays and scalar (https://github.com/goto40/msg-dsl-mixed-lookup-and-path):

```
//...
MsgStruct:
    'struct' name=ID '{'
        (msg_attributes+=MsgAttribute)*
    '}';

MsgAttribute: MsgScalarAttribute|MsgArrayAttribute;

MsgScalarAttribute: 'scalar' name=ID ':' msg_type=[MsgType:FQN];
MsgArrayAttribute: 'array' name=ID ':' msg_type=[MsgType:FQN] '[' dim=[MsgArrayDim:FQN] ']';
//...
```

Generating code from the list of attributes of a struct 
implies distinguishing the type of attribute (e.g. `MsgScalarAttribute`
or `MsgArrayAttribute`):

When this is done with the generated `isMsgArrayAttribute`
and `isMsgScalarAttribute` I found that (obviously) the `MsgArrayAttribute`
is-a `MsgScalarAttribute`... That confused me (is it a structural type check?).

The following code "works":
```
            if (isMsgArrayAttribute(attr)) {
              // process array attribute
            }
            else if (isMsgScalarAttribute(attr)) {
              // process scalar attribute
            }
```

The following code generates a problem for the else path:
```
            if (isMsgScalarAttribute(attr)) {
              // process scalar attribute
              // process array attribute (is-a MsgScalarAttribute???)
            }
            else if (isMsgArrayAttribute(attr)) {
              // never happens (impossible)
            }
```

Is this the intended way? Should I check for the `$type`-field to identify the underlying model type?

**Comment by msujew** [comment, marked as answer]
Hey @goto40,

I'm a bit confused, since I couldn't reproduce the problem that you experience with your repository. In general, the `langium-cli` will generate code (the `AstReflection`) that allows to perform Class-like inheritance checking based on the `$type` attribute and nothing else.

I've used a small test script:

```js
const ast = require('./language-server/generated/ast');
const isScalar = ast.isMsgScalarAttribute({ $type: 'MsgScalarAttribute' });
const isArray = ast.isMsgScalarAttribute({ $type: 'MsgArrayAttribute' });
console.log('scalar: ' + isScalar);
console.log('array: ' + isArray);
```

However, it prints that `MsgArrayAttribute` isn't a `MsgScalarAttribute` as expected. Can you provide me with a test case where we can clearly see the unexpected behavior?

**Reply by goto40** [reply]
You are right. I was confused by my vscode indicating that `attr` will have type `never`.
![Screenshot from 2022-08-03 13-02-31](https://user-images.githubusercontent.com/16887130/182593505-4e5f5447-a981-4dea-96dc-12672a5d5749.png)


However, **compiling and executing the code generated behavior as expected!**

**Reply by msujew** [reply]
Interesting. I guess that TypeScript knows that `MsgScalarAttribute` is a structural subtype of `MsgArrayAttribute`, which confuses it, since that would mean that `isMsgScalarAttribute` has to return true. If it returns false, it assumes that it also can't be a `MsgArrayAttribute`, but rather `never`.
